src/java.base/share/classes/java/util/Arrays.java

```diff
@@ -3454,7 +3462,9 @@ public static void fill(float[] a, @IndexOrHigh({"#1"}) int fromIndex, @IndexOrH
      * @throws ArrayStoreException if the specified value is not of a
      *         runtime type that can be stored in the specified array
      */
-    public static void fill(@PolyInterned @PolyNull @PolySigned Object[] a, @PolyInterned @PolyNull @PolySigned Object val) {
+    @CFComment("Parameters are not polymorphic, as that would unsoundly allow, e.g. passing an array of non-nullable "
```

Comment:
Ah, because of array covariance? I guess I would be inclined to say that, if you leave array covariance on, then you can get yourself in trouble in your _own_ code without ever calling an API. If we were to try to annotate to avoid such problems, then we might have to restrict even further, like maybe making `Arrays.asList` reject nullable type arguments, lest someone call `Arrays.<@Nullable String>asList(nonnullStrings)` and then write a `null` into to. I guess my view is that the widening happens in the callers, so it's not up to the method signature to worry about what might have already happened? Given that, I'd stick with the `@PolyNull` annotations that `fill` had.

(Alternatively, if you don't buy that, I think we could still leave `@PolyNull Object[] a`, even if we don't keep `@PolyNull Object val`.)

Reply:
Thanks, I hadn't thought about `asList`. In https://github.com/eisop/jdk/pull/85/commits/26dfd78aa71311c7be2a7467ca54e6ae0038eb40 I've tried to improve the text and added a comment to `asList`.

To me `fill` is more dangerous: it directly modifies the argument, so the array covariance can lead to a modification that the caller might not expect.
For `asList`, I would hope most callers don't retain a reference to the argument, so if the list is updated later, nobody could observe the violation in the array.

I agree that we should in general annotate APIs with invariant arrays in mind. Let me know if this restriction on `fill` would cause many false positives in your setting.

Reply:
I've changed the `fill` signature back and just kept the comment. Covariant arrays should be disabled for soundness. Something better could be done here, but let's not do it now.

Reply:
I did a search a few days ago that showed a number of calls of the form `Collections.{shuffle,reverse}(Arrays.asList(array))`, so people definitely do some of that.

I'm happy to see the current signature remain at the moment.